### ISSUE TITLE
ap derived from chain is incompatible with handwritten ap

### ISSUE THREAD

**Issue by rpominov** [rpominov/basic-streams#19]
Fantasy Land and Static Land require the following:

> If a method can be derived and the type also has a hand written version of that method, derived and hand written methods must be equivalent. And if two derived version of a method can be created (for instance we can derive `map` using `ap` or `chain`), they must be equivalent.

Although I'm still not 100% sure specs should require that, see discussion in FL chat https://gitter.im/fantasyland/fantasy-land?at=570a28a4c65c9a6f7f284c1c

The problem is `ap` derived from Streams `chain` is pretty meaningless. It works something like this (I use `map2` instead to simplify example, but behavior is same with `ap`):

```
map2 derived from ap that derived from chain:

a    1----2----3-----
b     --1----2----3--
b'         --2----3--
b''            ---3--

a+b  ---2----4----6--
             4    6
                  6
```

It spawns a version of `b` at each event in `a`, and produces an event on each event from each spawned version of `b`, so first problem we get an increasing number of duplicates for each event. Secondly we don't get an update at events from `a` only at events from `b`. Compare to how hand written `ap` works:

```
map2 derived from hand written ap

a    1----2----3-----
b    ---1----2----3--

a+b  ---2-3--4-5--6--
```

We can partially solve this by using `chainLatest` as default `chain`, which will look something like this:

```
map2 derived from ap that derived from chainLatest:

a    1----2----3-----
b     --1--
b'         --2--
b''             --3--

a+b  ---2----4----6--
```

So we'll get rid of duplicates, but we'll still don't get updates on events from `a`, and we'll constantly resubscribe to `b` which is depending on how `b` implemented can be a huge problem. So this is not really a solution.

My current conclusion is that a Stream can be Monad or can be Applicative but not both at the same time. I will think more about it.

**Comment by rpominov** [comment]
Seems like the derived `ap` is lawful at least, I did some testing: https://github.com/rpominov/basic-streams/commit/3ccece78796becc62ea32ae8a993122c82cc5fce

It doesn't pass Applicative's interchange rewritten with `map` though, we just got lucky with how `of` works in Streams.

So not everything is too broken here, seems like we have more or less lawful and usable `Monad` with more or less lawful but not usable derived `Applicative`.

Looking forward to explore how things are in https://github.com/rpominov/basic-signals , it might be much better there.
